### PR TITLE
Feature branch

### PR DESCRIPTION
**Pull request by Ashah15**
Building of an exclusive clubhouse where registered members can write embarrassing posts about non-members. Inside the clubhouse, members can see who the author of a post is but, non-members can only see the story and wonder who wrote it.

### CODE REVIEW

File: members-only/config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Rails.application.routes.draw do
+
+  get  '/signup',  to: 'users#new'
```
Comment (bolah2009) [inline members-only/config/routes.rb]: There's no `new` action in the user controller, hence I can't signup as a new user.

image

File: members-only/.stickler.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    fixer: true
+fixers:
+    enable: true
+files:
+  ignore:
+    - "bin/*"
+    - "db/*"
+    - "config/*"
+    - "Guardfile"
+    - "Rakefile"
```
Comment (bolah2009) [inline members-only/.stickler.yml]: Let's add this stickler config file to the root of the project. Currently, stickler is not checking this PR.

**Review by bolah2009 — CHANGES_REQUESTED** [review]
Awesome job team!

Please check the comments below.

#### :recycle:  Make these changes and submit another code review request. :+1: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

File: members-only/app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+class User < ApplicationRecord
+	attr_accessor :remember_token
+	before_save { self.email = email.downcase }
+	before_create :remember_token
```
Comment (maelfosso) [inline members-only/app/models/user.rb]: Define a method that will set the `remember_token` attribute

File: members-only/app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+module SessionsHelper
+	def log_in(user)
+    session[:user_id] = user.id
+  end
```
Comment (maelfosso) [inline members-only/app/helpers/sessions_helper.rb]: Think about using `cookies`
Take a look on this please https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_method_requirement.png

File: members-only/app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  # Remembers a user in a persistent session.
+  def remember(user)
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
```
Comment (maelfosso) [inline members-only/app/helpers/sessions_helper.rb]: merge this with `login` method

File: members-only/app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user && user.authenticated?(:remember, cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
```
Comment (maelfosso) [inline members-only/app/helpers/sessions_helper.rb]: Too long
Just check whether or not the user with the cookies remember_token exists

File: members-only/app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  def logged_in?
+    !current_user.nil?
+  end
+
+  # Forgets a persistent session.
+  def forget(user)
+    user.forget
+    cookies.delete(:user_id)
+    cookies.delete(:remember_token)
+  end
```
Comment (maelfosso) [inline members-only/app/helpers/sessions_helper.rb]: Line `34` is not really needed.

File: members-only/app/views/posts/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+
+	<div class="row py-5">
+		
+
+   <div class="col-12">
+   	<h1 class="px-3">Secret Posts</h1>
+
+   <% if logged_in? %>
+    <p class="px-3 d-block"><%= @user.name %> is currently logged in</p>
```
Comment (maelfosso) [inline members-only/app/views/posts/index.html.erb]: I don't really understand why you need this.
What if you directly use `current_user`?

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required

Hello @Ashah15 @MarvellousUbani 

Thank you for that awesome introduction.
The project is well working but there are some changes to do concerning the code 

1. Please, forget about using `Session`. Only think on `cookies`

2. Also, read the comments below

Kindly make the changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: members-only/app/models/user.rb (hunk 3 above)
Comment (MarvellousUbani) [reply members-only/app/models/user.rb]: This is done on the user model file

File: members-only/app/helpers/sessions_helper.rb (hunk 4 above)
Comment (MarvellousUbani) [reply members-only/app/helpers/sessions_helper.rb]: We removed sessions and used cookies

File: members-only/app/helpers/sessions_helper.rb (hunk 5 above)
Comment (MarvellousUbani) [reply members-only/app/helpers/sessions_helper.rb]: We also did this as well

File: members-only/app/helpers/sessions_helper.rb (hunk 6 above)
Comment (MarvellousUbani) [reply members-only/app/helpers/sessions_helper.rb]: We made it a one liner, finding the user by the id in the cookie

File: members-only/app/helpers/sessions_helper.rb (hunk 7 above)
Comment (MarvellousUbani) [reply members-only/app/helpers/sessions_helper.rb]: Done.

File: members-only/app/views/posts/index.html.erb (hunk 8 above)
Comment (MarvellousUbani) [reply members-only/app/views/posts/index.html.erb]: We did this as well

File: members-only/app/models/user.rb (hunk 3 above)
Comment (maelfosso) [reply members-only/app/models/user.rb]: Where is, please?
There is still 
```ruby
before_create :remember_token
```
With an attribute having the same name 
```ruby
attr_accessor :reminder_token
```

File: members-only/app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  # Remembers a user in the database for use in persistent sessions.
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
+  end
+
+  # Returns true if the given token matches the digest.
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
```
Comment (maelfosso) [inline members-only/app/models/user.rb]: Why this line, please?

File: members-only/app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+  # Returns true if the user is logged in, false otherwise.
+  def logged_in?
+    !current_user.nil?
+  end
+
+
+  def log_out
+    current_user = nil
```
Comment (maelfosso) [inline members-only/app/helpers/sessions_helper.rb]: Don't you think that, here, it's more like
```ruby
@current_user = nil
```

File: members-only/app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
+module SessionsHelper
+	def log_in(user)
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
```
Comment (maelfosso) [inline members-only/app/helpers/sessions_helper.rb]: Please, at the end of the login, set the `@current_user` variable directly

File: members-only/app/models/user.rb (hunk 3 above)
Comment (MarvellousUbani) [reply members-only/app/models/user.rb]: We assumed this set the user remember token attribute 

> Remembers a user in the database for use in persistent sessions.
```
  def remember
    self.remember_token = User.new_token
    update_attribute(:remember_digest, User.digest(remember_token))
  end
```

And then whenever we used 

```
def log_in(user)
    user.remember
    cookies.permanent.signed[:user_id] = user.id
    cookies.permanent[:remember_token] = user.remember_token
  end
```
then the user called the remember method which is on the model.

Is this okay perhaps, or is there a better way to do it please

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required

Hello,

Really Nice improvement :clap: 
However, there are things to changes

1. According to that [description](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/current_user_requirement.png), you need to define a method **current_user=(user)**. Implementing that will bring some changes in your code.

2. Don't forget to read that https://github.com/Ashah15/members/pull/1#discussion_r362186637

3. Please read the comment below.

Kindly make the changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: members-only/app/models/user.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+                    uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, presence: true, length: { minimum: 6 }
+
+  # Returns the hash digest of the given string.
+  def User.digest(string)
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+                                                  BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
+  end
```
Comment (maelfosso) [inline members-only/app/models/user.rb]: Please, read that https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_create_requirement.png

**Review by maelfosso — COMMENTED** [review]
Please, also take a look at that.

File: app/helpers/sessions_helper.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+    @current_user = user
+  end
+
+
+  # Returns the user corresponding to the remember token cookie.
+  def current_user
+      @current_user = User.find_by(id: cookies.signed[:user_id])
+  end
```
Comment (ebukaume) [inline app/helpers/sessions_helper.rb:13]: [OPTIONAL]

Consider memoizing `@current_user` so you don't need to hit the database if @current had already been set. 

=> `@current_user ||= User.find_by(id: cookies.signed[:user_id])`

**Review by ebukaume — APPROVED** [review]
Nice work team. You have done well with this project.

**You can now go ahead and merge this project**.

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Pull request merged**